src/magic_folder/web.py

```diff
@@ -114,11 +134,158 @@ class APIv1(Resource, object):
+    # config...
+
+    def __attrs_post_init__(self):
+        Resource.__init__(self)
+
+    def render_GET(self, request):
+        """
+        List all participants of this folder
+        """
+        collective = participants_from_collective(
+            self._folder_config.collective_dircap,
+            self._folder_config.upload_dircap,
+            self._tahoe_client,
+        )
+        d = collective.list()
+
+        def listed(participants):
+            reply = {
+                part.name: {
+                    "personal_dmd": part.dircap.decode("ascii"),
+                    # not tracked properly yet
+                    # "public_key": part.verify_key.encode(Base32Encoder),
+                }
+                for part in participants
+            }
+            request.setResponseCode(http.OK)
+            _application_json(request)
+            request.write(json.dumps(reply).encode("utf8"))
+        d.addCallback(listed)
+
+        def failed(reason):
+            # probably should log this failure
+            request.setResponseCode(http.INTERNAL_SERVER_ERROR)
+            _application_json(request)
+            request.write(json.dumps({"reason": "unexpected error processing request"}))
+            return None
+        d.addErrback(failed)
+        d.addBoth(lambda ignored: request.finish())
+        return NOT_DONE_YET
+
+    def render_POST(self, request):
+        """
+        Add a new participant to this folder with details from the JSON-encoded body.
+        """
+        body = request.content.read()
+        try:
+            participant = json.loads(body)
+            required_keys = {
+                "author",
+                "personal_dmd",
+            }
+            required_author_keys = {
+                "name",
+                # not yet
+                # "public_key_base32",
+            }
+            if set(participant.keys()) != required_keys:
+                raise ValueError("Require input: {}".format(", ".join(sorted(required_keys))))
+            if set(participant["author"].keys()) != required_author_keys:
+                raise ValueError("'author' requires: {}".format(", ".join(sorted(required_author_keys))))
+
+            author = create_author(
+                participant["author"]["name"],
+                # we don't yet properly track keys but need one
+                # here .. this won't be correct, but we won't use
+                # it .. following code still only looks at the
+                # .name attribute
+                VerifyKey(os.urandom(32)),
+            )
+
+            dmd = tahoe_uri_from_string(participant["personal_dmd"])
+            if not IDirnodeURI.providedBy(dmd):
+                raise ValueError("personal_dmd must be a directory-capability")
+            if not dmd.is_readonly():
+                raise ValueError("personal_dmd must be read-only")
+            personal_dmd_cap = participant["personal_dmd"]
+        except ValueError as e:
```

Comment:
Does  `ValueError` get raised by anything other than our code? I wonder if if it would make sense to have a custom exception that gets handled here, instead of using `ValueError`.

Reply:
I *thought* I was copying that style from elsewhere, but now I don't see where .. I don't *think* that `ValueError` comes from any of the called functions .. but even if it did, returning that as an error still seems appropriate?

Reply:
It isn't clear that we want to expose the message of a random other `ValueError` unchanged to the user, though.

Reply:
yeah, fair point .. I'll make a custom exception